pv211_utils/evaluation_metrics.py

```diff
@@ -0,0 +1,338 @@
+
+    mp_score.value /= len(queries)
+
+    return mp_score.value
+
+
+def mean_recall(system: IRSystemBase, queries: OrderedDict,
+                judgements: Set[JudgementBase], k: int, num_processes: int) -> float:
+    """Evaluate system for given queries and judgements with mean recall metric.
+    Where first k documents will be used in evaluation.
+
+    Args:
+    ----
+    system : IRSystemBase
+        System to be evaluated.
+    queries : OrderedDict
+        Queries to be searched.
+    judgements : Set[JudgementBase]
+        Judgements.
+    k : int
+        Parameter defining evaluation depth.
+    num_processes : int
+        Parallelization parameter defining number of processes to be used to run the evaluation.
+
+    Returns:
+    -------
+    float
+        Mean recall score from interval [0, 1].
+    """
+    manager = Manager()
+    mr_score_lock = manager.Lock()
+    mr_score = manager.Value('f', 0.0)
+
+    worker_recall = partial(_calc_recall, system,
+                            _judgements_obj_to_id(judgements),
+                            k, mr_score_lock, mr_score)
+
+    process_pool = Pool(processes=num_processes)
+    process_pool.map(worker_recall, list(queries.values()))
+
+    mr_score.value /= len(queries)
+
+    return mr_score.value
+
+
+def normalized_discounted_cumulative_gain(system: IRSystemBase,
+                                          queries: OrderedDict,
+                                          judgements: Set[JudgementBase],
+                                          k: int, num_processes: int) -> float:
+    """Evaluate system for given queries and judgements with normalized
+    discounted cumulative gain metric. Where first k documents will be used in evaluation.
+
+    Args:
+    ----
+    system : IRSystemBase
+        System to be evaluated.
+    queries : OrderedDict
+        Queries to be searched.
+    judgements : Set[JudgementBase]
+        Judgements.
+    k : int
+        Parameter defining evaluation depth.
+    num_processes : int
+        Parallelization parameter defining number of processes to be used to run the evaluation.
+
+    Returns:
+    -------
+    float
+        Normalized discounted cumulative gain score from interval [0, 1].
+    """
+    manager = Manager()
+    ndcg_score_lock = manager.Lock()
+    ndcg_score = manager.Value('f', 0.0)
+
+    worker_ndcg = partial(_calc_ndcg, system,
+                          _judgements_obj_to_id(judgements),
+                          k, ndcg_score_lock, ndcg_score)
+
+    process_pool = Pool(processes=num_processes)
+    process_pool.map(worker_ndcg, list(queries.values()))
+
+    ndcg_score.value /= len(queries)
+
+    return ndcg_score.value
+
+
+def mean_bpref(system: IRSystemBase, queries: OrderedDict,
+               judgements: Set[JudgementBase], k: int, num_processes: int) -> float:
+    """Evaluate system for given queries and judgements with bpref metric.
+    Where first k documents will be used in evaluation.
```

Comment:
```suggestion
    """Evaluate system for given queries and judgements with bpref metric.
    Where first k documents will be used in evaluation.
    
    Notes
    -----
    Since bpref differentiates between unjudged and irrelevant documents,
    this class only computes bpref for datasets with complete relevant judgements
    that cover exhaustively all pairs of queries and documents. At the time of writing,
    this is only true of the Cranfield dataset. This can only be remediated by changing
    the JudgementBase class to differentiate between unjudged and irrelevant
    documents.
```

Reply:
should be done